Only sync addresses of the own ip family

uablrek opened the pull request:
For ipv4 only ipv4 addresses are synced on the kube-ipvs0 device
and the same for ipv6. Fixes #70113


**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

The sync function tries to sync all addresses regardless of family (ipv4 or ipv6). This caused failures and will also cause problem for the upcoming dual-stack.

This PR make the sync function only care about addresses of the own family on the `kube-ipvs0` device.

**Which issue(s) this PR fixes**:

Fixes #70113

**Special notes for your reviewer**:

Testing is done manually by adding addresses of the "other" family to `kube-ipvs0` and check that they are untouched and that no failures are logged.

Also the special case in #70113 is tested; bring `kube-ipvs0` and get an automatically assigned link-local ipv6 address. The log was no longer spammed with failure records.

For ipv6 there is a problem; the bug reported in #65006 prevents proper testing.

By temporary hard-code a fix for the problem as described in the issue above it is possible to test for ipv6. The function of leaving ipv4 addresses untoched works as well as removing manually added ipv6 addresses that should no be there works fine.

But removing the auto-generated ipv6 address does not work even in a ipv6-only cluster. It can be removed manually so it is not "forbidden" to remove it. I will look at this problem and add to this PR if I find anything, but I would like to get things going and this fixes the problem for ipv4 which is most users concern.

**Does this PR introduce a user-facing change?**:

NONE

```release-note
```

Commit:
Only sync addresses of the own ip family

For ipv4 only ipv4 addresses are synced on the kube-ipvs0 device
and the same for ipv6. Fixes #70113

uablrek commented:
/sig network

m1093782566 commented:
/ok-to-test

/release-note-none

m1093782566 commented:
/lgtm

/approve

Commit:
Applied gofmt -s -w

m1093782566 commented:
/retest

m1093782566 commented:
@uablrek 

Please take a look at the failed UT:

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //pkg/proxy/ipvs:go_default_test
-----------------------------------------------------------------------------
E1207 10:40:58.345053      15 ipset.go:172] vstring (7) is not a valid version string: illegal version string "7"
E1207 10:40:58.345873      15 ipset.go:172] vstring (total junk) is not a valid version string: could not parse "total junk" as version
E1207 10:40:58.368744      15 service.go:122] Service ns1/svc1 has no healthcheck nodeport
--- FAIL: TestCleanLegacyBindAddr (0.00s)
    proxier_test.go:2932: Expected number of remaining bound addrs after cleanup to be 2. Got 3
    proxier_test.go:2941: Expected remainingAddrsMap map[1002:ab8::2:1:true 1.2.3.4:true], got map[1.2.3.4:true 1002:ab8::2:1:true 1002:ab8::2:2:true]
FAIL

```

uablrek commented:
@m1093782566 First; sorry for not running UT locally before submitting. I will do that in the future.

The test assumes that the ipvs proxier can handle both ipv4 and ipv6 addresses at the same time. The
"addresses that were bound to ipvs0 in the latest sync loop"
includes both families, which is possible in UT but could never happen in real life with the current code.

For support of dual-stack there are 2 options as I see it;

* Extend the proxier instance to handle both families.
* Make 2 instances of the proxier, one for ipv4 and one for ipv6.

I have assumed the later and I think that is the best way but it is not my call.

BTW: I have cheched similar things in `kube-proxy` and a problem that will arise is that `ipset` are not divided on family as `iptables` so ipv6 and ipv4 sets share the same "namespace" so to speak. I used a name prefix in `kube-router`.

So, what I would like to do to fix the UT failure is to alter the test-case, and probably write a few new ones, and assume different proxier instances for ipv4 and ipv6.

But if the proxier should be extended to handle both families in the same instance then my PR should be revoked and I will solve the problem in some other way. The auto-generated link-local address should be possible to remove and that should also fix the problem.

And another thing; the `kube-ipvs0` device is "DOWN" in my system and then no link-local ipv6 address is generated. Do you have any idea why the `kube-ipvs0` device is  "UP" in some other systems?

m1093782566 commented:
@uablrek 

I also noticed that ipset does not support IPV6. For now, before K8S support dual-stack, I think we can safely assume kube-proxy only support IPV4 or IPV6 at the same time.

> And another thing; the kube-ipvs0 device is "DOWN" in my system and then no link-local ipv6 address is generated. Do you have any idea why the kube-ipvs0 device is "UP" in some other systems?

Unfortunately, I have no idea now :(

Commit:
Handle ipv4 OR ipv6 addresses in TestCleanLegacyBindAddr

An ipv6-only test added.

uablrek commented:
I have modified the failing test and added an ipv6-only test. UT pass, but with some printouts.

Should I take some action for the conflicting files?

m1093782566 commented:
I just merged other PR which also modify proxier.go and proxier_test.go.

I think you need to rebase to the HEAD.

uablrek commented:
My updates in pkg/proxy/ipvs/proxier_test.go are obsolete but I can't find any way to "undo" them :confused:.

I end up with conflicts in rebase no matter what, even if I revert the commit.

The UT passes with my PR in proxy.go (after resolving the conflict).

I will do a new PR with only the proxy.go updates based on the current master and close this PR.

uablrek commented:
I made a new PR; https://github.com/kubernetes/kubernetes/pull/71911

One should probably rebase immediately when a conflict occurs before pushing new stuff.
